File: python/memory_sync/restore.py
"""Restore a session: pull, rewrite origin->target, write to target project dir."""
import os
import shutil

from . import config as config_mod
from . import encoding
from . import manifest as manifest_mod
from . import rewriter
from . import store as store_mod


def restore(checkpoint_id: str, target_cwd: str, cfg_path: str) -> str:
    cfg = config_mod.load(cfg_path)
    work_dir = os.path.join(os.path.expanduser("~"), ".memory-sync", "work", cfg.project_id)
    s = store_mod.GitStore(cfg.store_url, work_dir)
    bundle = s.get(checkpoint_id)

    with open(os.path.join(bundle, "manifest.json")) as f:
        m = manifest_mod.load(f.read())
    uuid = m.session_uuid
    target_project = encoding.project_dir(target_cwd)
    os.makedirs(target_project, exist_ok=True)

    # transcript: rewrite origin -> target
    src_jsonl = os.path.join(bundle, uuid + ".jsonl")
    if os.path.isfile(src_jsonl):
        dst_jsonl = os.path.join(target_project, uuid + ".jsonl")
        rewriter.rewrite_session(src_jsonl, dst_jsonl, m.origin_root, target_cwd)

    # memory: copy as-is (paths rewritten in bodies is M3's concern; M2 copies raw)
    src_mem = os.path.join(bundle, "memory")
    if os.path.isdir(src_mem):
        dst_mem = os.path.join(target_project, "memory")
        if os.path.exists(dst_mem):
            shutil.rmtree(dst_mem)
        shutil.copytree(src_mem, dst_mem)

    return uuid

File: go/memorysync/restore.go
package memorysync

import (
	"compress/gzip"
	"fmt"
	"io"
	"os"
	"path/filepath"
)

// Restore pulls the bundle, rewrites origin→target via RewriteSession,
// writes to the target project dir, returns the session UUID.
func Restore(checkpointID, targetCwd, cfgPath string) (string, error) {
	cfg, err := LoadConfig(cfgPath)
	if err != nil {
		return "", fmt.Errorf("load config: %w", err)
	}
	home, err := os.UserHomeDir()
	if err != nil {
		return "", fmt.Errorf("get home dir: %w", err)
	}
	workDir := filepath.Join(home, ".memory-sync", "work", cfg.ProjectID)
	store, err := NewGitStore(cfg.Store.URL, workDir)
	if err != nil {
		return "", fmt.Errorf("new git store: %w", err)
	}
	bundle, err := store.Get(checkpointID)
	if err != nil {
		return "", fmt.Errorf("store get: %w", err)
	}

	// read manifest
	mData, err := os.ReadFile(filepath.Join(bundle, "manifest.json"))
	if err != nil {
		return "", fmt.Errorf("read manifest: %w", err)
	}
	m, err := Load(string(mData))
	if err != nil {
		return "", fmt.Errorf("parse manifest: %w", err)
	}
	uuid := m.SessionUUID

	targetProject := ProjectDir(targetCwd)
	if err := os.MkdirAll(targetProject, 0755); err != nil {
		return "", fmt.Errorf("mkdir target project: %w", err)
	}

	// transcript: gunzip if compressed, then RewriteSession (atomic)
	dstJSONL := filepath.Join(targetProject, uuid+".jsonl")
	tmpJSONL := dstJSONL + ".tmp"
	var srcTranscript string
	if m.TranscriptCompressed {
		srcGz := filepath.Join(bundle, uuid+".jsonl.gz")
		srcTranscript = dstJSONL + ".decompressed"
		if err := gunzipFile(srcGz, srcTranscript); err != nil {
			return "", fmt.Errorf("gunzip transcript: %w", err)
		}
		defer os.RemoveAll(srcTranscript)
	} else {
		srcTranscript = filepath.Join(bundle, uuid+".jsonl")
	}
	if _, err := os.Stat(srcTranscript); err == nil {
		if _, err := RewriteSession(srcTranscript, tmpJSONL, m.OriginRoot, targetCwd); err != nil {
			return "", fmt.Errorf("rewrite session: %w", err)
		}
		if err := os.Rename(tmpJSONL, dstJSONL); err != nil {
			return "", fmt.Errorf("rename transcript: %w", err)
		}
	}
	// skipped files warning
	if len(m.SkippedFiles) > 0 {
		fmt.Fprintf(os.Stderr, "warning: %d tool-results file(s) were skipped during checkpoint (too large): %v\n", len(m.SkippedFiles), m.SkippedFiles)
	}

	// memory: copy as-is (NOT path-rewritten — spec §13)
	srcMem := filepath.Join(bundle, "memory")
	if _, err := os.Stat(srcMem); err == nil {
		dstMem := filepath.Join(targetProject, "memory")
		if err := os.RemoveAll(dstMem); err != nil {
			return "", fmt.Errorf("remove old memory: %w", err)
		}
		if err := copyDir(srcMem, dstMem); err != nil {
			return "", fmt.Errorf("copy memory: %w", err)
		}
	}

	return uuid, nil
}

func gunzipFile(src, dst string) error {
	in, err := os.Open(src)
	if err != nil {
		return err
	}
	defer in.Close()
	gz, err := gzip.NewReader(in)
	if err != nil {
		return err
	}
	defer gz.Close()
	out, err := os.Create(dst)
	if err != nil {
		return err
	}
	defer out.Close()
	_, err = io.Copy(out, gz)
	return err
}
